kernel/comps/virtio/src/device/input/device.rs

```diff
@@ -181,44 +185,133 @@ impl InputDevice {
         size
     }
 
-    /// Negotiate features for the device specified bits 0~23
-    pub(crate) fn negotiate_features(features: u64) -> u64 {
-        assert_eq!(features, 0);
-        0
-    }
-}
-
-impl aster_input::InputDevice for InputDevice {
-    fn handle_irq(&self) -> Option<()> {
-        // one interrupt may contains serval input, so it should loop
-        loop {
-            let Some(event) = self.pop_pending_event() else {
-                return Some(());
+    fn handle_irq(&self) {
+        // Returns ture if there may be more events to handle
+        let handle_event = |event: &EventBuf| -> bool {
+            let event: VirtioInputEvent = {
+                let mut reader = event.reader();
+                reader.read_val()
             };
+
             match event.event_type {
-                0 => return Some(()),
+                0 => return false,
                 // Keyboard
                 1 => {}
                 // TODO: Support mouse device.
-                _ => continue,
+                _ => return true,
             }
+
             let status = match event.value {
                 1 => KeyStatus::Pressed,
                 0 => KeyStatus::Released,
-                _ => return Some(()),
+                _ => return false,
             };
+
             let event = InputEvent::KeyBoard(Key::try_from(event.code).unwrap(), status);
             info!("Input Event:{:?}", event);
 
             let callbacks = self.callbacks.lock();
             for callback in callbacks.iter() {
-                callback.call((event,));
+                callback(event);
             }
+
+            true
+        };
+
+        // one interrupt may contains serval input, so it should loop
+        while self.pop_pending_event(&handle_event) {}
+    }
+
+    /// Negotiate features for the device specified bits 0~23
+    pub(crate) fn negotiate_features(features: u64) -> u64 {
+        assert_eq!(features, 0);
+        0
+    }
+}
+
+/// A event table consists of many event buffers,
+/// each of which is large enough to contain a `VirtioInputEvent`.
+#[derive(Debug)]
+struct EventTable {
+    stream: DmaStream,
+    num_events: usize,
+}
+
+impl EventTable {
+    fn new(num_events: usize) -> Self {
+        assert!(num_events * mem::size_of::<VirtioInputEvent>() <= PAGE_SIZE);
+
+        let vm_segment = {
+            let mut options = VmAllocOptions::new(1);
+            options.is_contiguous(true);
+            options.alloc_contiguous().unwrap()
+        };
+
+        let default_event = VirtioInputEvent::default();
+        for idx in 0..num_events {
+            let offset = idx * EVENT_SIZE;
+            vm_segment.write_val(offset, &default_event).unwrap();
```

Comment:
Writing a value repeatedly seems to be a common need. Consider adding a new method called `writes_vals`.

```rust
impl VmIo {
    /// Write a sequence of values given by an iterator (`iter`) from the specified offset (`offset`).
    /// 
    /// The write process stops until the VM object does not have enough remaining space
    /// or the iterator returns `None`. If any value is written, the function returns `Ok(nr_written)`,
    /// where `nr_written` is the number of the written values.
    ///
    /// The offset of every value written by this method is aligned to the `align`-byte boundary.
    /// Naturally, when `align` equals to `0` or `1`, then the argument takes no effect:
    /// the values will be written in the most compact way.
    ///
    /// # Example
    ///
    /// Initializing an VM object with the same value can be done easily with `write_values`.
    ///
    /// ```
    /// use core::iter::self;
    ///
    /// let _nr_values = vm_obj.write_values(0, iter::repeat(0_u32), 0).unwrap();
    /// ```
    fn write_values<T: Pod, I: Iterator<Item = T>>(&self, offset: usize, iter: I, align: usize) -> Result<usize> {
        // ...
    }
}
```

This extension to `VmIo` should be added in a separate commit.

(I am not quite sure that the trait bound for `I` is the most appropriate one)

Reply:
I have rewritten the signature as
```rust
fn write_vals<'a, T: Pod + 'a, I: Iterator<Item = &'a T>>(
        &self,
        offset: usize,
        iter: I,
        align: usize,
    ) -> Result<usize> 
```

Ideally, the `Item` of `Iterator` can be either `T` or `&T`. But I can't find a type to support both. So I use `&T` here.